quickwit/quickwit-control-plane/src/indexing_scheduler/scheduling/scheduling_logic.rs

```diff
@@ -0,0 +1,620 @@
+}
+
+fn assert_enforce_nodes_max_load_post_condition(
+    problem: &SchedulingProblem,
+    node_assignment: &NodeAssignment,
+) {
+    let total_load = node_assignment.total_load(problem);
+    assert!(total_load <= problem.node_max_load(node_assignment.node_id));
+}
+
+// ----------------------------------------------------
+// Phase 3
+// Place unassigned sources.
+//
+// We use a greedy algorithm as a simple heuristic here.
+// We go through the sources in decreasing order of their load.
+//
+// We then try to place as many shards as possible in the node with the
+// highest load.
+
+fn place_unassigned_shards(
+    problem: &SchedulingProblem,
+    solution: &mut SchedulingSolution,
+) -> BTreeMap<SourceOrd, u32> {
+    let mut unassigned_shards: Vec<Source> = compute_unassigned_sources(problem, solution);
+    unassigned_shards.sort_by_key(|source| {
+        let load = source.num_shards * source.load_per_shard.get();
+        Reverse(load)
+    });
+    let mut node_with_least_loads: BinaryHeap<(Load, NodeOrd)> =
+        compute_node_available_capacity(problem, solution);
+    let mut unassignable_shards = BTreeMap::new();
+    for source in unassigned_shards {
+        let num_shards_unassigned =
+            place_unassigned_shards_single_source(&source, &mut node_with_least_loads, solution);
+        // We haven't been able to place this source entirely.
+        if num_shards_unassigned != 0 {
+            unassignable_shards.insert(source.source_id, num_shards_unassigned);
+        }
+    }
+    assert_place_unassigned_shards_post_condition(problem, solution, &unassignable_shards);
+    unassignable_shards
+}
+
+fn assert_place_unassigned_shards_post_condition(
+    problem: &SchedulingProblem,
+    solution: &SchedulingSolution,
+    unassigned_shards: &BTreeMap<SourceOrd, u32>,
+) {
+    // We make sure we all shard are cound as place or unassigned.
+    for source in problem.sources() {
+        let num_assigned_shards: u32 = solution
+            .node_assignments
+            .iter()
+            .map(|node_assignment| node_assignment.num_shards(source.source_id))
+            .sum();
+        assert_eq!(
+            num_assigned_shards
+                + unassigned_shards
+                    .get(&source.source_id)
+                    .copied()
+                    .unwrap_or(0),
+            source.num_shards
+        );
+    }
+    // We make sure that all unassigned shard cannot be placed.
+    for node_assignment in &solution.node_assignments {
+        let available_capacity: Load = node_assignment.node_available_capacity(problem);
+        for (&source_id, &num_shards) in unassigned_shards {
+            assert!(num_shards > 0);
+            let source = problem.source(source_id);
+            assert!(source.load_per_shard.get() > available_capacity);
+        }
+    }
+}
+
+fn place_unassigned_shards_single_source(
+    source: &Source,
+    node_with_least_loads: &mut BinaryHeap<(Load, NodeOrd)>,
```

Comment:
nitpick: `node_with_least_loads` is built from `compute_node_available_capacity`. There is a confusion about what's inside the binary heap: is is the load or is the available capacity. Maybe rename the variable to `node_available_capacities`.